### PR TITLE
[Java] CRC for Archive

### LINKED ISSUE #353: [Archiver] Add CRC as option, per fragment, use session id to store CRC
Note: This is a place holder item, which may or may not be worth doing.
This is on write to file. This will produce overheads as currently we do not parse fragments on write. The overwritter sessionId is not used on read since the fragment header is replaced on replay.

### PR DESCRIPTION
**Pull request by vyazelenko**
Support for CRC in Archive consists of the following parts:
- Ability to compute CRC checksums when writing fragments into segment files (configured via `Archive.Context#recordingCrcEnabled()`)
- Ability to verifying CRC checksum during replay (configured via `Archive.Context#replayCrcEnabled()`)
- Ability to verify CRC checksums using `ArchiveTool#verify` and `ArchiveTool#verifyRecording` methods
- Ability to compute and persist CRC checksums using `ArchiveTool#checksum` and `ArchiveTool#checksumRecording` methods

Closes #353

### CODE REVIEW

File: aeron-archive/src/main/java/io/aeron/archive/Archive.java [hunk 1]
```diff
@@ -369,6 +381,38 @@ public static Archive launch(final Context ctx)
...
+        /**
+         * Do not compute the CRC for each recorded fragment by default.
+         */
+        public static final boolean RECORDING_CRC_ENABLED_DEFAULT = false;
+
+        /**
+         * Property that specifies fully qualified class name of the {@link io.aeron.archive.checksum.Checksum}
+         * to be used for CRC checksum computation during recording.
+         */
+        public static final String RECORDING_CRC_CHECKSUM_PROP_NAME = "aeron.archive.recording.crc.checksum";
```
Comment (mjpt777) [inline aeron-archive/src/main/java/io/aeron/archive/Archive.java]: Pairs should be recording and replaying, or record and replay. Action vs activity. We should pick one or the the other.

`aeron.archive.recording.checksum` and `aeron.archive.replaying.checksum`, or `aeron.archive.record.checksum` and aeron.archive.replay.checksum`. Either can work and stays abstract.  Simply setting either can be used to indicate if enabled or not, no need for a separate property.

File: aeron-archive/src/main/java/io/aeron/archive/ReplaySession.java [hunk 2]
```diff
@@ -338,8 +353,12 @@ private int replay() throws IOException
                 result = publication.tryClaim(dataLength, bufferClaim);
                 if (result > 0)
                 {
+                    if (crcEnabled)
+                    {
+                        applyCrc(frameOffset, alignedLength);
```
Comment (mjpt777) [inline aeron-archive/src/main/java/io/aeron/archive/ReplaySession.java]: Apply before claim to keep window as small as possible to reduce cache misses and probability of buffer unblocking.

File: aeron-archive/src/main/java/io/aeron/archive/ReplaySession.java [hunk 3]
```diff
@@ -504,7 +536,7 @@ static boolean isInvalidHeader(
     {
         return
             DataHeaderFlyweight.termOffset(buffer, 0) != termOffset ||
-            DataHeaderFlyweight.termId(buffer, 0) != termId ||
-            DataHeaderFlyweight.streamId(buffer, 0) != streamId;
+                DataHeaderFlyweight.termId(buffer, 0) != termId ||
```
Comment (mjpt777) [inline aeron-archive/src/main/java/io/aeron/archive/ReplaySession.java]: Don't further indent.

File: aeron-samples/src/main/java/io/aeron/samples/archive/EmbeddedRecordingThroughput.java [hunk 4]
```diff
@@ -116,7 +117,7 @@ public long streamMessagesForRecording()
                 idleStrategy.idle();
             }
 
-            final long startMs = System.currentTimeMillis();
+            final long start = System.nanoTime();
```
Comment (mjpt777) [inline aeron-samples/src/main/java/io/aeron/samples/archive/EmbeddedRecordingThroughput.java]: Indicate time unit in name.

File: aeron-samples/src/main/java/io/aeron/samples/archive/EmbeddedReplayThroughput.java [hunk 5]
```diff
@@ -80,11 +81,11 @@ public static void main(final String[] args) throws Exception
             do
             {
                 System.out.printf("Replaying %,d messages%n", NUMBER_OF_MESSAGES);
-                final long start = System.currentTimeMillis();
+                final long start = System.nanoTime();
```
Comment (mjpt777) [inline aeron-samples/src/main/java/io/aeron/samples/archive/EmbeddedReplayThroughput.java]: Indicate time unit in name.

**Comment by mjpt777** [comment]
Generally looks really good and with a few tweaks we can merge.

File: aeron-archive/src/main/java/io/aeron/archive/RecordingWriter.java [hunk 6]
```diff
@@ -126,6 +147,21 @@ else if (isPaddingFrame)
         }
     }
 
+    private void computeCRC(final UnsafeBuffer termBuffer, final int termOffset, final int length)
```
Comment (mjpt777) [inline aeron-archive/src/main/java/io/aeron/archive/RecordingWriter.java]: Abbreviations should be camel case.

File: aeron-archive/src/main/java/io/aeron/archive/Archive.java (hunk 1 above)
Comment (vyazelenko) [reply aeron-archive/src/main/java/io/aeron/archive/Archive.java]: I went with `aeron.archive.record.checksum` and `aeron.archive.replay.checksum`.

File: aeron-archive/src/main/java/io/aeron/archive/ReplaySession.java (hunk 2 above)
Comment (vyazelenko) [reply aeron-archive/src/main/java/io/aeron/archive/ReplaySession.java]: Done

File: aeron-archive/src/main/java/io/aeron/archive/ReplaySession.java (hunk 3 above)
Comment (vyazelenko) [reply aeron-archive/src/main/java/io/aeron/archive/ReplaySession.java]: Intellij auto-format...

File: aeron-samples/src/main/java/io/aeron/samples/archive/EmbeddedRecordingThroughput.java (hunk 4 above)
Comment (vyazelenko) [reply aeron-samples/src/main/java/io/aeron/samples/archive/EmbeddedRecordingThroughput.java]: done

File: aeron-samples/src/main/java/io/aeron/samples/archive/EmbeddedReplayThroughput.java (hunk 5 above)
Comment (vyazelenko) [reply aeron-samples/src/main/java/io/aeron/samples/archive/EmbeddedReplayThroughput.java]: done

File: aeron-archive/src/main/java/io/aeron/archive/RecordingWriter.java (hunk 6 above)
Comment (vyazelenko) [reply aeron-archive/src/main/java/io/aeron/archive/RecordingWriter.java]: done

**Pull request merged**